### PR TITLE
Initial HLint work

### PR DESCRIPTION
**Pull request by galderz**
Initial HLint work fixing warnings.

Error free `hlint` output can be achieved via:

```
hlint --cpp-define=__GLASGOW_HASKELL__=800 --cpp-define=x86_64_HOST_ARCH=1 --cpp-define=mingw32_HOST_OS=1 .
```

Still TODO:
- [x] Add documentation on HLint.
- [x] Move "How To Contribute" section to docs section and add HLint there. 
- [ ] Link from README to contributing guide.
- [x] Deal with suggestions.
- [ ] Add to CI.

### CODE REVIEW

**Comment by galderz** [comment]
I've just updated the PR to fix the error that CI had found, add more ignores and finally add suggestions. Assuming CI passes, only docu would be left to do.

File: Foundation/Monad.hs [hunk 1]
```diff
@@ -1,7 +1,6 @@
 {-# LANGUAGE CPP #-}
 {-# LANGUAGE DeriveGeneric #-}
 {-# LANGUAGE DeriveDataTypeable #-}
-{-# LANGUAGE StandaloneDeriving #-}
```
Comment (NicolasDP) [inline Foundation/Monad.hs]: I think this one is needed for older compilers (ghc.7.8 and 7.10).

see: https://travis-ci.org/haskell-foundation/foundation/jobs/241461362#L300

Comment (galderz) [inline Foundation/Monad.hs]: Yup, CI spotted it. Just fixed it.

File: Foundation/Array/Unboxed.hs [hunk 2]
```diff
@@ -1020,7 +1019,7 @@ reverse a
         loop !i
             | i == end  = return ()
             | otherwise = primMbaWrite ma i (primBaIndex ba (sizeAsOffset (endI - i))) >> loop (i+Offset 1)
-    goAddr !end !ma !(Ptr ba) !srcStart = loop (Offset 0)
+    goAddr !end !ma (Ptr ba) !srcStart = loop (Offset 0)
```
Comment (ndmitchell) [inline Foundation/Array/Unboxed.hs]: Should this have the ! on `ba`?

File: Foundation/Array/Unboxed/Mutable.hs [hunk 3]
```diff
@@ -161,18 +161,18 @@ new sz
 {-# INLINE new #-}
 
 mutableSame :: MUArray ty st -> MUArray ty st -> Bool
-mutableSame (MUVecMA sa ea _ ma) (MUVecMA sb eb _ mb) = and [ sa == sb, ea == eb, bool# (sameMutableByteArray# ma mb)]
-mutableSame (MUVecAddr s1 e1 f1) (MUVecAddr s2 e2 f2) = and [ s1 == s2, e1 == e2, finalPtrSameMemory f1 f2 ]
-mutableSame (MUVecMA {})     (MUVecAddr {})   = False
-mutableSame (MUVecAddr {})   (MUVecMA {})     = False
+mutableSame (MUVecMA sa ea _ ma) (MUVecMA sb eb _ mb) = (sa == sb) && (ea == eb) && bool# (sameMutableByteArray# ma mb)
```
Comment (ndmitchell) [inline Foundation/Array/Unboxed/Mutable.hs]: I wouldn't usually bother with the brackets around `sa == sb`

File: Foundation/Check/Main.hs [hunk 4]
```diff
@@ -67,11 +67,11 @@ filterTestMatching cfg testRoot
...
     | otherwise                = testFilter [] testRoot
   where
-    match acc s = or $ fmap (flip isInfixOf currentTestName) $ testNameMatch cfg
+    match acc s = or (flip isInfixOf currentTestName <$> testNameMatch cfg)
       where currentTestName = fqTestName (s:acc)
     or [] = False
     or (x:xs)
-        | x == True = True
+        | x         = True
         | otherwise = or xs
```
Comment (ndmitchell) [inline Foundation/Check/Main.hs]: Isn't this just absolutely standard `or` from the base library? Why replicate it?

**Comment by ndmitchell** [comment]
This should probably be enabled on the CI as well - see https://github.com/ndmitchell/hlint#running-with-continuous-integration

**Comment by galderz** [comment]
> This should probably be enabled on the CI as well - see https://github.com/ndmitchell/hlint#running-with-continuous-integration

Certainly. I'll update the TODO list at the top.

File: Foundation/Array/Unboxed.hs (hunk 2 above)
Comment (galderz) [inline Foundation/Array/Unboxed.hs]: Hmmmm, dunno... @vincenthz ?

File: Foundation/Array/Unboxed/Mutable.hs (hunk 3 above)
Comment (galderz) [inline Foundation/Array/Unboxed/Mutable.hs]: Ah, true... From chat w/ @vincenthz yday, at times he likes having parenthesis to read the code easier. I'll let him decide on this one.

**Comment by NicolasDP** [comment]
Good idea about adding hlint in our CI too. About Travis, I would rather like we only run the hlint in a different job. A job dedicated for running hlint.

**Comment by ndmitchell** [comment]
How come a dedicated job? Should only add a few seconds to the time. When you add weeder (I hope!) would you also put that in a dedicated job?

**Comment by galderz** [comment]
Yeah, I'll check with the guys later on how to add a job that runs first with just `HLint`.

Indeed, planning to add Weeder too. I'll create a separate issue for this.

**Comment by galderz** [comment]
Btw, I'm keeping an eye on build results, fixing stuff and pushing new changes... that's why CI keeps rebuilding 😎

**Comment by NicolasDP** [comment]
I think it would make things easier to spot on travis-ci.

* one job per compiler;
* one job for hlint;
* one job for weeder and detecting dead code ? Why not...
* one job for doctest ?

The idea is to be able to quickly see what the problems are by just looking at the job list for the associated PR. Avoiding to spend too much time scrolling the logs to find that it's not the compilation or the tests who failed but a linting issue or a documentation issue...

This is open to discussion anyway. I'm not going to make the decision for the whole community.

**Comment by ndmitchell** [comment]
That kind of makes sense - apart from doctest - you really want to run the doctests in every compiler since they might operate differently. Things like HLint and Weeder I can see the reasoning.

**Comment by galderz** [comment]
Added another commit with documentation and moved the contributing guide to docs.

File: Foundation/Check/Main.hs (hunk 4 above)
Comment (galderz) [inline Foundation/Check/Main.hs]: I've created #316

**Pull request merged**